src/simulator/api.rs

```diff
@@ -0,0 +1,245 @@
+            }
+        }
+        // In a real implementation, this should update all NodeSelector
+    }
+    fn add_getters(&mut self, getters: Vec<Channel<Getter>>) {
+        for getter in getters {
+            let previous = self.getters.insert(
+                getter.id.clone(),
+                GetterWithState {
+                    getter:getter,
+                    state: None
+                });
+            assert!(previous.is_none());
+        }
+        // In a real implementation, this should update all GetterSelectors
+    }
+    fn add_setters(&mut self, setters: Vec<Channel<Setter>>)  {
+        for setter in setters {
+            let previous = self.setters.insert(setter.id.clone(), setter);
+            assert!(previous.is_none());
+        }
+        // In a real implementation, this should update all SetterSelectors
+    }
+
+    fn add_watch(&mut self, options: Vec<WatchOptions>, cb: Box<Fn(WatchEvent)>) {
+        let cb = Arc::new(cb);
+        for opt in options {
+            self.watchers.push((opt, cb.clone()));
+        }
+    }
+
+    fn inject_getter_value(&mut self, id: Id<Getter>, value: Value) {
+        let mut getter = self.getters.get_mut(&id).unwrap();
+        getter.set_state(value.clone());
+
+        // The list of watchers watching for new values on this getter.
+        let watchers = self.watchers.iter().filter(|&&(ref options, _)| {
+            options.should_watch_values &&
+                options.source.matches(&getter.getter)
+        });
+        for watcher in watchers {
+            watcher.1(WatchEvent::Value {
+                from: id.clone(),
+                value: value.clone()
+            });
+        }
+    }
+
+    fn put_value(&mut self,
+                 selectors: Vec<SetterSelector>,
+                 value: Value,
+                 cb: Box<Fn(Vec<(Id<Setter>, Result<(), Error>)>)>)
+    {
+        // Very suboptimal implementation.
+        let setters = self.setters
+            .values()
+            .filter(|setter|
+                    selectors.iter()
+                    .find(|selector| selector.matches(setter))
+                    .is_some());
+        let results = setters.map(|setter| {
+            let result;
+            let internal_result;
+            if value.get_type() == setter.mechanism.kind.get_type() {
+                result = Ok(());
+                internal_result = Ok(());
+            } else {
+                result = Err(foxbox_taxonomy::api::Error::TypeError);
+                internal_result = Err(format!("Invalid type, expected {:?}, got {:?}", value.get_type(), setter.mechanism.kind.get_type()));
+            }
+            (*self.post_updates)(Update::Put {
+                id: setter.id.clone(),
+                value: value.clone(),
+                result: internal_result
+            });
+            (setter.id.clone(), result)
+        }).collect();
+        cb(results)
+    }
+}
+
+#[derive(Clone)]
+pub struct APIFrontEnd {
```

Comment:
Could you document this?

Also, I'm not sure that `APIFrontEnd` is a very good name. Maybe `DevicesSimulation`?